ext/standard/file.c

```diff
@@ -455,132 +455,121 @@ PHP_FUNCTION(file_get_contents)
 /* {{{ Write/Create a file with contents data and return the number of bytes written */
 PHP_FUNCTION(file_put_contents)
 {
-	php_stream *stream;
-	char *filename;
-	size_t filename_len;
-	zval *data;
+	php_stream *stream = NULL;
 	ssize_t numbytes = 0;
+	zend_string *filename;
+	zval *data;
 	zend_long flags = 0;
 	zval *zcontext = NULL;
 	php_stream_context *context = NULL;
-	php_stream *srcstream = NULL;
+	php_stream *data_stream = NULL;
+	zend_string *data_str = NULL;
 	char mode[3] = "wb";
 
 	ZEND_PARSE_PARAMETERS_START(2, 4)
-		Z_PARAM_PATH(filename, filename_len)
+		Z_PARAM_PATH_STR(filename)
 		Z_PARAM_ZVAL(data)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_LONG(flags)
 		Z_PARAM_RESOURCE_OR_NULL(zcontext)
 	ZEND_PARSE_PARAMETERS_END();
 
+	/* Type checks for data */
 	if (Z_TYPE_P(data) == IS_RESOURCE) {
-		php_stream_from_zval(srcstream, data);
+		/* This throws AND returns if resource is not a stream */
+		php_stream_from_zval(data_stream, data);
+	} else if (Z_TYPE_P(data) == IS_ARRAY) {
+		/* Check array is an array of strings */
+		if (zend_hash_num_elements(Z_ARRVAL_P(data))) {
+			smart_str smart_str = {0};
+			zval *tmp;
+
+			ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(data), tmp) {
+				zend_string *tmp_entry_str = NULL;
+				zend_string *entry_str = zval_try_get_tmp_string(tmp, &tmp_entry_str);
+
+				if (UNEXPECTED(entry_str == NULL)) {
+					smart_str_free(&smart_str);
+					RETURN_THROWS();
+				}
+				if (ZSTR_LEN(entry_str)) {
+					// Append to smart string
+					smart_str_append(&smart_str, entry_str);
```

Comment:
So previously we wrote it directly to the streams and now smart_str is used. It seems to me that this can result in more to be used at a single point which might not be ideal especially if you have got lots of items in the array. This seems like a potential regression to me.